### ISSUE TITLE
nonzero_compared_to_zero: Lint comparison of `NonZero` values with zero

### ISSUE THREAD

**Issue by sgued** [rust-lang/rust-clippy#12238]
### What it does

Lint comparison of `NonZero` values with `0`.

### Advantage

Such a comparison will always have a constant value.

### Example

```rust
let a = NonZeroU32::new(3).unwrap();
if a.get() == 0 {
    println!("This is unreachable");
}
```

Could be written as:

```rust
let a = NonZeroU32::new(3).unwrap();
```

**Comment by Centri3** [comment]
Makes sense to me. We could generalize this to any struct with the `#[rustc_layout_scalar_valid_range_start]`/`#[rustc_layout_scalar_valid_range_end]` attributes

**Comment by KisaragiEffective** [comment]
How are there types with those range attributes? I can't think there are other than non-zero friends...

**Comment by Centri3** [comment]
I wasn't really thinking about any crate in specific, or inheriting. Just the struct that had them applied specifically.

It'd be better than checking for `NonZeroU8`, `NonZeroI8`, etc., tho now I realize that was switched to just `NonZero<T>` so there's no difference at all

> How are there types with those range attributes? I can't think there are other than non-zero friends...

There are some in libstd but afaik all of them are either internal or just can't reasonably be compared against smth outside their range, e.g., `BorrowedFd`. You'd need to convert to `RawFd` first to check if it's -1

**Comment by Kixunil** [comment]
Could also include `enum Foo { Bar = 0, Baz = 1 } let x = Foo::Bar; if x as u8 == 42 { ... }`
